pkg/query-service/app/http_handler.go

```diff
@@ -707,6 +742,25 @@ func (aH *APIHandler) deleteDashboard(w http.ResponseWriter, r *http.Request) {
 
 }
 
+func (aH *APIHandler) queryDashboardVars(w http.ResponseWriter, r *http.Request) {
+
+	query := r.URL.Query().Get("query")
+	if query == "" {
+		respondError(w, &model.ApiError{Typ: model.ErrorBadData, Err: fmt.Errorf("query is required")}, nil)
+		return
+	}
+	if strings.Contains(strings.ToLower(query), "alter table") {
+		respondError(w, &model.ApiError{Typ: model.ErrorBadData, Err: fmt.Errorf("query shouldn't alter data")}, nil)
+		return
+	}
+	dashboardVars, err := (*aH.reader).QueryDashboardVars(r.Context(), query)
```

Comment:
My thought process is this interface methods shouldn't bother if it's used in the context of API or an another internal method call. They do some specific thing and return an error to the caller and it's up to the caller to handle. If it's an API handler that will in turn returns `ApiError` or an internal method will have different mechanism. In an ideal this error would have return internal error but since client wanted to handle it gracefully we changed it bad request.

Reply:
hmm..so it basically boils down to: who should decide the status code sent back, the caller or the called? How is the caller in a better position to decide the status code? It will again have to check the type of error returned in order to provide correct status code. I have found it easier to return the type of error (status code being a type here) to give the caller direct actionables. As there can be many nested function calls, each layer can either respect the returned type of error from the called function or can choose to override it too. I would suggest to checkout some more examples on this. Would be an interesting exploration.

Reply:
My point the concept of status code shouldn't be inside these interface methods. The errors are usually propagated up the call stack and status code is decided at the time of handler responding -- I find that idiomatic. Sure, I can link some relevant examples.

Reply:
Fex examples from jaeger query service.

metrics store reader [interface](https://github.com/jaegertracing/jaeger/blob/71f2bb99f65b502ec935b05adbd6f8025c8f3bdf/storage/metricsstore/interface.go#L25) and [implementation](https://github.com/jaegertracing/jaeger/blob/7baadbd748294574451817197f83dc9bbd1a2e38/plugin/metrics/prometheus/metricsstore/reader.go#L98-L161)

trace store reader [interface](https://github.com/jaegertracing/jaeger/blob/71f2bb99f65b502ec935b05adbd6f8025c8f3bdf/storage/spanstore/interface.go#L35) and [implementation](https://github.com/jaegertracing/jaeger/blob/main/cmd/query/app/querysvc/query_service.go#L66-L119)